iam-token-manager/v1.ts

```diff
@@ -0,0 +1,190 @@
+ */
+
+import extend = require('extend');
+import { sendRequest } from '../lib/requestwrapper';
+
+export type Options = {
+  iamApikey?: string;
+  iamAccessToken?: string;
+  iamUrl?: string;
+}
+
+export interface IamTokenData {
+  access_token: string;
+  refresh_token: string;
+  token_type: string;
+  expires_in: number;
+  expiration: number;
+}
+
+export class IamTokenManagerV1 {
+  name: string;
+  serviceVersion: string;
+  protected url: string;
+  protected tokenInfo: IamTokenData;
+  private iamApikey: string;
+  private userAccessToken: string;
+
+  /**
+   * IAM Token Manager Service
+   *
+   * Retreives, stores, and refreshes IAM tokens.
+   *
+   * @param {Object} options
+   * @param {String} options.iamApikey
+   * @param {String} options.iamAccessToken
+   * @param {String} options.iamUrl - url of the iam api to retrieve tokens from
+   * @constructor
+   */
+  constructor(options: Options) {
+    this.url = options.iamUrl || 'https://iam.ng.bluemix.net/identity/token';
+    this.tokenInfo = {} as IamTokenData;
+    if (options.iamApikey) {
+      this.iamApikey = options.iamApikey;
+    }
+    if (options.iamAccessToken) {
+      this.userAccessToken = options.iamAccessToken;
+    }
+  }
+
+  /**
+   * This function sends an access token back through a callback. The source of the token
+   * is determined by the following logic:
+   * 1. If user provides their own managed access token, assume it is valid and send it
+   * 2. If this class is managing tokens and does not yet have one, make a request for one
+   * 3. If this class is managing tokens and the token has expired, refresh it
+   * 4. If this class is managing tokens and has a valid token stored, send it
+   *
+   * @param {Function} cb - callback function that the token will be passed to
+   */
+  public getToken(cb: Function) {
+    if (this.userAccessToken) {
+      // 1. use user-managed token
+      return cb(null, this.userAccessToken);
+    } else if (!this.tokenInfo.access_token) {
+      // 2. request an initial token
+      this.requestToken((err, tokenResponse) => {
+        this.saveTokenInfo(tokenResponse);
+        return cb(err, this.tokenInfo.access_token);
+      });
+    } else if (this.isTokenExpired()) {
+      // 3. refresh a token
+      this.refreshToken((err, tokenResponse) => {
+        this.saveTokenInfo(tokenResponse);
+        return cb(err, this.tokenInfo.access_token);
+      });
+    } else {
+      // 4. use valid managed token
+      return cb(null, this.tokenInfo.access_token);
+    }
+  }
+
+  /**
+   * Request an IAM token using an API key.
+   *
+   * @param {Function} cb - The callback that handles the response.
+   * @returns {void}
+   */
+  public requestToken(cb: Function): void {
+    const parameters = {
+      options: {
+        url: this.url,
+        method: 'POST',
+        headers: {
+          'Content-type': 'application/x-www-form-urlencoded',
+          Authorization: 'Basic Yng6Yng='
```

Comment:
Could this cause any issues since it's hardcoded? I know you mentioned that it acts like a secret -- you need to pass the same authorization header when refreshing a token.

Since the SDK should be managing any token that is requested/refreshed (and is ignoring the case where users provide a refresh key and secret...at least for now), it seems safe to use this hardcoded value. But just want to check.

Reply:
It should not cause any issues in the code. The only downside is the use case where a user is using this token manager class on it's own - they would not be able to specify a custom authorization header. Everything should still work, they just would be stuck with the auth header that we provide.